src/NServiceBus.Core/Unicast/Messages/MessageMetadataRegistry.cs

```diff
@@ -48,26 +49,41 @@ public MessageMetadata GetMessageMetadata(string messageTypeIdentifier)
         {
             Guard.AgainstNullAndEmpty(nameof(messageTypeIdentifier), messageTypeIdentifier);
 
-            var messageType = GetType(messageTypeIdentifier);
+            var cacheHit = cachedTypes.TryGetValue(messageTypeIdentifier, out var messageType);
 
-            if (messageType == null)
+            if (!cacheHit)
             {
-                Logger.DebugFormat("Message type: '{0}' could not be determined by a 'Type.GetType', scanning known messages for a match", messageTypeIdentifier);
+                messageType = GetType(messageTypeIdentifier);
 
-                foreach (var item in messages.Values)
+                if (messageType != null)
+                {
+                    cachedTypes[messageTypeIdentifier] = messageType;
+                }
+                else
                 {
-                    var messageTypeFullName = GetMessageTypeNameWithoutAssembly(messageTypeIdentifier);
+                    Logger.DebugFormat("Message type: '{0}' could not be determined by a 'Type.GetType', scanning known messages for a match", messageTypeIdentifier);
```

Comment:
should we change that message slightly towards something like `Message type: '{0}' could not be determined by the type identifier '{1}', scanning known messages for a match` or something like that?

Reply:
fine with leaving as is. In case of dynamic loading being disabled, it would still log `could not be determined by a 'Type.GetType'` which is technically not correct. But it's more of a minor thing.